### PR TITLE
Feature/animation system

### PR DESCRIPTION
**Pull request by yalania**
-  Some bugs from Resource Management fixed
- Added Skinning shader
- Added animation component and controller
- Fixed animation and skeleton resources

### CODE REVIEW

File: Engine/Animation/AnimController.cpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	if (channel_found)
+	{
+		float delta = current_sample - current_keyframe;
+		pos = Utils::Interpolate(current_translation, next_translation, delta);
+	}
+
+	return channel_found;
+}
+
+bool AnimController::GetRotation(const std::string& channel_name, Quat & rot)
```
Comment (JorxSS) [inline Engine/Animation/AnimController.cpp]: Code style: Quat& and also change rot for rotation?(I know can be undestood already but.. code style and readability)

File: Engine/Animation/AnimController.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+	if (channel_found)
+	{
+		float delta = current_sample - current_keyframe;
+		rot = Utils::Interpolate(current_rotation, next_rotation, delta);
+	}
+
+	return channel_found;
+}
+
+bool AnimController::GetTransformation(const std::string& channel_name, float4x4 & transform)
```
Comment (JorxSS) [inline Engine/Animation/AnimController.cpp]: float4x4&

File: Engine/Animation/AnimController.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+		}
+		else
+		{
+			playing = false;
+		}
+	}
+
+}
+
+bool AnimController::GetTranslation(const std::string& channel_name, float3 & pos)
```
Comment (JorxSS) [inline Engine/Animation/AnimController.cpp]: float3&

File: Engine/Animation/AnimController.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+	if (channel_found)
+	{
+		float delta = current_sample - current_keyframe;
+		transform = Utils::Interpolate(current_global_tranform, next_global_tranform, delta);
+	}
+	return channel_found;
+}
+
+void AnimController::GetChannelGlobalTransform(size_t channel_index,const Animation::Channel & channel, const std::vector<Animation::Channel> & channels, math::float4x4 &current_global_tranform)
```
Comment (JorxSS) [inline Engine/Animation/AnimController.cpp]: const Animation::Channel& channel, std::vector<Animation::Channel>& channels, math::float4x4&

File: Engine/Animation/AnimController.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+	bool GetTranslation(const std::string& channel_name, float3& pos);
+	bool GetRotation(const std::string& channel_name, Quat& rot);
+	bool GetTransformation(const std::string& channel_name, float4x4 & transform);
+
+	void GetChannelGlobalTransform(size_t channel_index, const Animation::Channel & channel, const std::vector<Animation::Channel> & channels, math::float4x4 &current_global_tranform);
+
+public:
+	std::shared_ptr<Animation> anim = nullptr;
+	std::shared_ptr<Skeleton> sk = nullptr;
```
Comment (JorxSS) [inline Engine/Animation/AnimController.h]: skeleton instead of sk?

File: Engine/Animation/AnimController.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	~AnimController() = default;
+	void Init();
+
+	void Play();
+	void Stop();
+	void Update(); 
+
+	bool GetTranslation(const std::string& channel_name, float3& pos);
+	bool GetRotation(const std::string& channel_name, Quat& rot);
+	bool GetTransformation(const std::string& channel_name, float4x4 & transform);
```
Comment (JorxSS) [inline Engine/Animation/AnimController.h]: float4x4&

File: Engine/Animation/AnimController.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+	void GetChannelGlobalTransform(size_t channel_index, const Animation::Channel & channel, const std::vector<Animation::Channel> & channels, math::float4x4 &current_global_tranform);
```
Comment (JorxSS) [inline Engine/Animation/AnimController.h]: same as comented  in the cpp file

File: Engine/Component/ComponentAnimation.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+	std::string animation_path;
+	config.GetString("AnimationResource", animation_path, "");
+	SetAnimation(App->resources->Load<Animation>(animation_path));
+
+	std::string skeleton_path;
+	config.GetString("SkeletonResource", skeleton_path, "");
+	SetSkeleton(App->resources->Load<Skeleton>(skeleton_path));
+}
+
+void ComponentAnimation::SetAnimation(std::shared_ptr<Animation> & animation)
```
Comment (JorxSS) [inline Engine/Component/ComponentAnimation.cpp]: code style, sorry you are going to hate me but.. blame on @mariofv and @Dartemiss

File: Engine/Component/ComponentAnimation.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+{
+	animation_controller->anim = animation;
+}
+void ComponentAnimation::SetSkeleton(std::shared_ptr<Skeleton> & skeleton)
```
Comment (JorxSS) [inline Engine/Component/ComponentAnimation.cpp]: code style here

File: Engine/Component/ComponentAnimation.h [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+#include "Component.h"
+
+#include "EditorUI/Panel/InspectorSubpanel/PanelComponent.h"
+
+#include <GL/glew.h>
+#include <memory>
+class AnimController;
+class GameObject;
+class Animation;
```
Comment (JorxSS) [inline Engine/Component/ComponentAnimation.h]: hateeeee me anabel... alphabetical order

File: Engine/Component/ComponentLight.h [hunk 11]
```diff
@@ -27,9 +27,8 @@ class ComponentLight : public Component
 		public: 
 			void ChangePointLightAttenuationValues(float range)
 			{
-				float augmented_range = 10 * range;
-				linear = 4.5 / (10 * augmented_range);
-				quadratic = 75.0 / (augmented_range*augmented_range);
+				linear = 4.5 / (10 * range);
+				quadratic = 75.0 / (range * range);
```
Comment (JorxSS) [inline Engine/Component/ComponentLight.h:31]: why this values? maybe make them as a macro?

File: Engine/Helper/Utils.h [hunk 12]
```diff
@@ -25,6 +24,10 @@ class Utils
 	void SaveFileContent(const std::string & source, std::string & destination);
 
 	void ReplaceStringInPlace(std::string & subject, const std::string & search, const std::string & replace);
+
+	static float3 Interpolate(const float3 & first, const float3 & second, float lambda);
+	static Quat Interpolate(const Quat & first, const Quat & second, float lambda);
+	static float4x4 Interpolate(const float4x4& first, const float4x4& second, float lambda);
```
Comment (JorxSS) [inline Engine/Helper/Utils.h]: code style here

File: Engine/Main/Globals.h [hunk 13]
```diff
@@ -69,14 +69,14 @@ enum class Fonts
...
-# define PRIMITIVE_TORUS_PATH "Resources/Primitives/Torus/Torus.ol"
-# define PRIMITIVE_QUAD_PATH "Resources/Primitives/Quad/Quad.ol"
+# define VIDEO_BILLBOARD_TEXTURE_PATH "Library/Metadata/35/3575636683"
+# define LIGHT_BILLBOARD_TEXTURE_PATH "Library/Metadata/26/2664975493"
+
+# define PRIMITIVE_CUBE_PATH "Library/Metadata/41/4135282941"
+# define PRIMITIVE_CYLINDER_PATH "Library/Metadata/40/4057255088"
+# define PRIMITIVE_SPHERE_PATH "Library/Metadata/35/3553469253"
+# define PRIMITIVE_TORUS_PATH "Library/Metadata/22/2202705559"
+# define PRIMITIVE_QUAD_PATH "Library/Metadata/35/3548710651"
```
Comment (JorxSS) [inline Engine/Main/Globals.h:79]: what are this numbers? those paths are always fix?

File: Engine/Module/ModuleDebugDraw.h [hunk 14]
```diff
@@ -4,11 +4,13 @@
 #include "Module.h"
 
 #include <GL/glew.h>
+#include <MathGeoLib.h>
 
 class Billboard;
 class ComponentCamera;
 class Grid;
 class IDebugDrawOpenGLImplementation;
+class GameObject;
```
Comment (JorxSS) [inline Engine/Module/ModuleDebugDraw.h]: alphabetical order

File: Engine/Module/ModuleResourceManager.h [hunk 15]
```diff
@@ -39,6 +39,7 @@ class ModuleResourceManager : public Module
 	ImportResult Import(const File& file, bool force = false);
 	void ImportAllFilesInDirectory(const File& file, bool force);
 	void CreatePrefab(const std::string &path, GameObject * gameobject_to_save) const;
+	uint32_t LoadCubemap(const std::vector<std::string> & faces_paths);
```
Comment (JorxSS) [inline Engine/Module/ModuleResourceManager.h]: cs here

File: Engine/ResourceManagement/Importer/ModelImporters/SkeletonImporter.cpp [hunk 16]
```diff
@@ -4,43 +4,55 @@
 
 #include "Main/Application.h"
 #include "Module/ModuleFileSystem.h"
-bool SkeletonImporter::ImportSkeleton(const aiScene* scene, const aiMesh* mesh, const std::string& imported_file, std::string& exported_file) const
+ImportResult SkeletonImporter::ImportSkeleton(const aiScene* scene, const aiMesh* mesh, const std::string& imported_file, float unit_scale_factor, Skeleton & skeleton) const
```
Comment (JorxSS) [inline Engine/ResourceManagement/Importer/ModelImporters/SkeletonImporter.cpp:7]: cs here

File: Game/resources/Shaders/shaders.json [hunk 17]
```diff
@@ -1,70 +1,76 @@
...
+    {
+      "Name": "Grid",
+      "Vertex": "Resources/Shaders/grid.vs",
+      "Fragment": "Resources/Shaders/grid.fs",
+      "Selectable": false
+    },
+    {
+      "Name": "Skinning",
+      "Vertex": "Resources/Shaders/skinning.vs",
+      "Fragment": "Resources/Shaders/blinn_phong.fs",
```
Comment (JorxSS) [inline Game/resources/Shaders/shaders.json:72]: the fs will be always blinn phong for skinning?

**Review by JorxSS — APPROVED** [review]

**Comment by JorxSS** [comment]
Amazing job. Sorry for the Code Style comments.  
Maybe a little guide in the PR to put another animation for testing would be nice.

File: Engine/Component/ComponentLight.h (hunk 11 above)
Comment (yalania) [reply Engine/Component/ComponentLight.h:31]: I don't know what is this @mariofv

File: Engine/Main/Globals.h (hunk 13 above)
Comment (yalania) [reply Engine/Main/Globals.h:79]: yes, now the meta files and the uuid of each file are very important and permanent. So in this should be fixed

**Review by RickyVimon — APPROVED** [review]
All fine

File: Engine/Component/ComponentLight.h (hunk 11 above)
Comment (mariofv) [reply Engine/Component/ComponentLight.h:31]: Yeah, a macro would be the correct implementation. These numbers are defined [here](http://wiki.ogre3d.org/Light+Attenuation+Shortcut).

**Pull request merged**